src/engine/renderer/tr_shade.cpp

```diff
@@ -1075,7 +1194,8 @@ static void Render_lightMapping( shaderStage_t *pStage )
 
 	gl_lightMappingShader->SetUniform_Color( color );
 
-	if ( tess.bspSurface && !enableLightMapping )
+	// It looks like WrapAroundLighting is meant for models.
+	if ( !tess.bspSurface && enableGridLighting )
```

Comment:
`!tess.bspSurface` is true when the surface comes from a model (not a bsp surface), if wrapAroundLighting is for models, then testing for `tess.bspSurface` was a bug.

I see no reason why world surfaces would be rendered with tricks.

I find no documentation about what is “wrap around lighting” for a surface renderer. Searching on the internet only gives me some kind of lamp and photography techniques involving lamps. So, in that context, “wrap around lighting” is about lamps, not about surface, and a shader processing surface pixels can do nothing about it.

The only thing I barely found last week was something about emulating the way light navigates within some materials (like skin) , and that is something a shader can help to implement, but if that is what we have, it's for models.

Reply:
Something good to know is that this code was likely written by me in a previous refactoring, same for the GLSL `#ifelse`.

Reply:
Well similarly to half-Lambert lighting, "wraparound lighting" seems to be a hack for making stuff less dark. With deluxe maps the (really stupid?) assumption is made that all light comes from a single direction. In reality there are often various sources, plus even if the light comes from a point source it can diffract and bounce. Therefore the hacks are needed to avoid stuff not facing towards the light direction being overly dark. "Wraparound lighting" artificially boosts the dot product of the normal and light direction, as if the light turns slightly to hit the surface, hence the "wrap around" name. Like a crude model of diffraction I guess.

Half-Lambert lighting does a different calculation to mess with the dot product of normal and light direction, so the two techniques should not be used at once.

It's true that these techniques make more sense for models than BSP surfaces. I assume that when there are multiple light sources, the map compiler would appropriately adjust the lightmaps and deluxemaps to get the desired final amount of illumination.

I suppose it's likely that wraparound lighting was some one-off Xreal test or something, so no game models using it exist.

Reply:
> It's true that these techniques make more sense for models than BSP surfaces. I assume that when there are multiple light sources, the map compiler would appropriately adjust the lightmaps and deluxemaps to get the desired final amount of illumination.

Especially, q3map2 knows how to bounce the light. This misses other mechanism that exists in real life, but at least in a room, something not facing towards the light direction gets lit by the light bouncing on other surfaces around (something that may be implementable in q3map2 to make it more realistic would be to read the specular map to bounce less when the surface is less shiny, it already bounces according to the color though).

Reply:
I just thought about something, maybe rim lighting can be useful on world surfaces when using dynamic lights?

Reply:
The code was NUKed:

- https://github.com/DaemonEngine/Daemon/pull/1015